scheds/rust/scx_rustlite/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,404 @@
+	__uint(key, 0);
+	__type(value, struct dispatched_task_ctx);
+	__uint(max_entries, MAX_ENQUEUED_TASKS);
+} dispatched SEC(".maps");
+
+/*
+ * Heartbeat timer used to periodically trigger the check to run the user-space
+ * scheduler.
+ *
+ * Without this timer we may starve the scheduler if the system is completely
+ * idle and hit the watchdog that would auto-kill this scheduler.
+ */
+struct usersched_timer {
+	struct bpf_timer timer;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, u32);
+	__type(value, struct usersched_timer);
+} usersched_timer SEC(".maps");
+
+/*
+ * Return true if the target task @p is a kernel thread.
+ */
+static inline bool is_kthread(const struct task_struct *p)
+{
+	return !!(p->flags & PF_KTHREAD);
+}
+
+/*
+ * Return true if the target task @p is the user-space scheduler.
+ */
+static inline bool is_usersched_task(const struct task_struct *p)
+{
+	return p->pid == usersched_pid;
+}
+
+/*
+ * Dispatch a task on the local per-CPU FIFO.
+ */
+static void dispatch_task_local(struct task_struct *p, u64 enq_flags)
+{
+	dbg_msg("dispatch (local): pid=%d", p->pid);
+	scx_bpf_dispatch(p, SCX_DSQ_LOCAL, slice_ns, enq_flags);
+	__sync_fetch_and_add(&nr_kernel_dispatches, 1);
+}
+
+/*
+ * Dispatch a task on the global FIFO.
+ */
+static void
+dispatch_task_global(struct task_struct *p, u64 payload, u64 enq_flags)
+{
+	/*
+	 * Tasks without a payload are dispatched directly by the kernel, tasks
+	 * with payload are dispatched by the user-space scheduler.
+	 */
+	if (!payload)
+		dbg_msg("dispatch (global): pid=%d", p->pid);
+	else
+		dbg_msg("dispatch (user): pid=%d payload=%llu", p->pid, payload);
+	scx_bpf_dispatch(p, SCX_DSQ_GLOBAL, slice_ns, enq_flags);
+	__sync_fetch_and_add(&nr_user_dispatches, 1);
+}
+
+/*
+ * Select the target CPU for a task that is ready to run.
+ */
+s32 BPF_STRUCT_OPS(rustlite_select_cpu, struct task_struct *p, s32 prev_cpu,
+		   u64 wake_flags)
+{
+	/*
+	 * Let's keep it simple for now and always return the previously used
+	 * cpu, .enqueue() will be called on this cpu and if it's idle we can
+	 * immediately dispatch the task bypassing the user-space scheduler.
+	 */
+	return prev_cpu;
+}
+
+/*
+ * Return true if the selected CPU for the task is immediately avaliable
+ * (user-space scheduler not required), false otherwise (user-space scheduler
+ * required).
+ */
+static bool is_task_cpu_available(struct task_struct *p)
+{
+	/*
+	 * Always dispatch per-CPU kthread on the same CPU, bypassing the
+	 * user-space scheduler (in this way we can to prioritize critical
+	 * kernel threads that may potentially slow down the entire system if
+	 * they are blocked for too long).
+	 */
+        if (is_kthread(p) && p->nr_cpus_allowed == 1)
+		return true;
+	/*
+	 * Keep using the same CPU if the task can only run there or if the CPU
+	 * is idle.
+	 */
+	if (p->nr_cpus_allowed == 1 ||
+	    scx_bpf_test_and_clear_cpu_idle(bpf_get_smp_processor_id()))
```

Comment:
So the whole workflow with which actual physical core `ops.enqueue()` is called on is pretty confusing and probably one of the leakier parts of the API, so I understand where the confusion is coming from here. Taking a step back, let me try to clarify some of the operations going on here:

`ops.select_cpu()` and `ops.enqueue()` are called from the CPU where `try_to_wake_up()` is being invoked in the kernel. That does not need to be the same CPU where the task that is currently being woken up was previously running on, however. For example, you could have something like this:
```
CPU 0                 CPU 1
task_x blocks

                       try_to_wake_up(task_x)
                       cpu = ops.select_cpu(task_x, 0); // returns 0, i.e. prev_cpu
                       if (cpu != prev_cpu)
                              /*
                               * Even if the cpu returned by ops.select_cpu()
                               * was different than prev_cpu, the task isn't
                               * actually migrated here. We're just updating
                               * @p's CPU and rq pointers, which as we'll see
                               * below, is how we determine which SCX_DSQ_LOCAL
                               * DSQ to dispatch the task to from ops.enqueue().
                               */
                               set_task_cpu(p, cpu);
                       activate_task(rq /* rq of CPU 1 */, p, en_flags);
                                ops.enqueue(p, en_flags);
                                        // 1 == bpf_get_smp_processor_id()
                                        scx_bpf_test_and_clear_cpu_idle(1);
```

So the problem with doing things this way is that `scx_bpf_test_and_clear_cpu_idle(bpf_get_smp_processor_id()))` will be the CPU that _called_ `try_to_wake_up()`. _Not_ the CPU returned by `ops.select_cpu()`.

It's admittedly very confusing because conceptually `ops.enqueue()` is called on the CPU that's returned by `ops.select_cpu()`. We can do this because in `scx_bpf_dispatch(p, SCX_DSQ_LOCAL, slice_ns, enq_flags);`, we enqueue the task on the local DSQ according to the _task's_ rq, _not_ the rq of the current core.

So, going back to the code here, I believe what we're trying to do is to directly dispatch the task if its previous CPU is still idle. If that's the case, I _think_ we should be able to just change `scx_bpf_test_and_clear_cpu_idle(bpf_get_smp_processor_id())` to `scx_bpf_test_and_clear_cpu_idle(p->thread_info.cpu)`. Or, the more common approach would probably be to call `scx_bpf_test_and_clear_cpu_idle(prev_cpu)` in `ops.select_cpu()`, and then use task local storage to mark that the task should enqueue directly on the LOCAL_DSQ.

Note that calling `scx_bpf_test_and_clear_cpu_idle(p->thread_info.cpu)` from `is_task_cpu_available()` may confuse some readers, as the core is only marked as idle in the internal idle mask tracking once there are no tasks left to run on the CPU. So even if `@p` is the only task running on `p->thread_info.cpu`, `scx_bpf_test_and_clear_cpu_idle(p->thread_info.cpu)` won't return `true` after it's gone through its first slice.

@htejun This seems to be a common point of confusion, so I think we should probably try to figure out a way to more clearly document the semantics.

Reply:
There's `scx_bpf_task_cpu()` helper to obtain the cpu associated with a task, and yeah, this can be confusing and needs better documentation. I was wrong before too in saying that `.select_cpu()`'s only role is optimizing hot-migration overhead. It determines the CPU `.enqueue()` can directly dispatch to.

As for the single thread left on the CPU case, because we put the previous task after dispatching the next one and automatically keeps running the previous if nothing was dispatched unless SCX_OPS_ENQ_LAST is set, that'd probably be okay behavior-wise, I think. ie. the `.enqueue()` for the current task is only going to run if `.dispatch()` dispatched another task beforehand.

Also, not a blocker but if the scheduler is going to do direct dispatches from the enqueue path, it may as well do full idle CPU picking in `.select_cpu()`? If you go that route, as this does represent a non-trivial loss of control on the userspace scheduler's part, it may make sense to make the behavior optional.

Reply:
Thanks @Decave and @htejun for all the clarifications on `.select_cpu()`, `.enqueue()` and the like, you both have been super helpful! Now it makes a lot more sense to me why other schedulers are using the per-task `force_local` flag.

For this scheduler at the end I think it would probably make sense to mimic what `scx_userland` is doing (so full idle cpu picking in `.select_cpu()`, and add some comments in the code to clarify the difference between the "current cpu" and tasks's currrent cpu.

But in perspective also having full idle CPU picking in `.select_cpu()` optional is an interesting idea and have like 1)  cpu selection performed by the dispatcher vs 2) cpu selection performed by the scheduler. 

The former would do idle picking in `.select_cpu()` and then we dispatch to the local DSQ when `force_local` is set, as usual. With the latter `.select_cpu()` can always return `prev_cpu`, then the user-space scheduler could access that cpu via `struct queued_task_ctx` and potentially override this selection based on some internal logic, then send back to the dispatcher the selected cpu for the task via `struct dispatched_task`, and if it's different, the dispatcher can then use `scx_bpf_dispatch()` with `SCX_DSQ_LOCAL_ON | cpu` in `ops.dispatch()`.

Does this make sense? Do you think it's doable or there's still something that I'm missing?

Reply:
In 2), one thing to consider is that `LOCAL_ON` dispatching is always the highest priority. Taking a step back, this goes into the whether the userspace scheduler is involved in deciding which CPU a given task ends up executing on. If we go there, the BPF part probably should also expose the idlemask and the current tasks to the userland scheduler so that it can make a more intelligent decision and we also likely would want to keep a tight leash on how many tasks are queued in the dispatcher as it's easier to get ordering inversions otherwise.